### ISSUE TITLE
rollup.watch stops watching after first change.

### ISSUE THREAD

**Issue by zwhitchcox** [rollup/rollup#1666]
https://github.com/zwhitchcox/rollup-starter/tree/7cdc2b7b340ef0c920e527cb703059762fb0385c

Clone the repo above, and `yarn install` and `yarn start`, and you'll see that watch doesn't work after the first go round.

Video of bug: https://youtu.be/T1JVpFWSkPQ

**Comment by zwhitchcox** [comment]
It actually does the same thing on https://github.com/rollup/rollup-starter-project too

**Comment by aubergene** [comment]
I just tried to recreate this with rollup version 0.51.8 and it seemed ok. Is it still an issue for you?

**Comment by tcoopman** [comment]
I just tested this with 0.51.8 and I have a similar issue with my repo (https://github.com/tcoopman/bucklescript-tea-starter-kit). It doesn't happen after the second time, but after a few times rollup always stops detecting changes for me.
I also had this with different repo's.

**Comment by ezze** [comment]
@zwhitchcox, @tcoopman, what are your IDEs? This happens when IDE is creating temporary files on source changes and [doesn't write to existing files directly](https://github.com/rollup/rollup/issues/1669#issuecomment-350248142).

**Comment by tcoopman** [comment]
I'm using visual studio code. I have no issues with any other watchers like
for example webpack.
For me it doesn't always stop after the first watch, most of the time it
stops later

**Comment by zwhitchcox** [comment]
I'm using vim

**Comment by max-l** [comment]
I experienced this issue using Pycharm, disabling the "safe writes" Pycharm settings solved the problem

**Comment by zwhitchcox** [comment]
This is still a problem for me, and I have verified that it is actually writing to the disk

**Comment by zwhitchcox** [comment]
Also, other watchers work for me, like facebook jest, and gaze. It's just rollup that doesn't work

**Comment by Tanoemon** [comment]
This happens to me while running browser-sync in parallel. So I think this is because of #1938 for me.

**Comment by Aldlevine** [comment]
I think I have the same issue here. I'm not sure if this is the case for everyone else, but in my scenario it will correctly run the first time any file is saved.

For example, if you have 3 files in the project and you start `rollup --watch`, saving "file1.js" will trigger the build. Saving "file1.js" again won't do anything, but saving "file2.js" will work correctly. Finally, saving both "file1.js" or "file2.js" will do nothing, but saving "file3.js" will work correctly.

Lastly, when I disable chokidar in the config file, everything seems to work as expected (see below). So I suspect this issue has something to do with the chokidar configuration. Either way, this seems like a successful workaround.

```javascript
// rollup.config.js
export default {
  ...
  watch: {
    chokidar: false
  },
  ...
}
```

**Comment by shellscape** [comment]
@zwhitchcox if you have a spare moment, could you confirm that the latest version exhibits the same behavior?

**Comment by zwhitchcox** [comment]
I actually can't, as I'm not using rollup anymore.. will close though

**Comment by angryziber** [comment]
The issue is still present with rollup 1.21.4 and Intellij IDEA.
Adding `chokidar: false` helps.

**Comment by paco3346** [comment]
I'm finding the opposite to be true, I have to use chokidar with `usePolling: true` to be a universal solution.

A lot of this boils down to how your watchers are set up and whether you're using atomic writes (safe writes in IntelliJ products). If you watch the whole directory things work fine because when the editor unlinks and adds a new file the watcher gets the notification. When you start to watch a specific file and it's atomically swapped you don't get the alert because the file's inode changed.

**Comment by paco3346** [comment]
After 2 full days on research and hacking at the source locally here's what I've concluded:

Rollup watches individual files instead of directories. This make for efficient watching [but breaks with atomic writes](https://nodejs.org/docs/latest/api/fs.html#fs_inodes) (safe writes) in modern editors. (@BobbyBabes mentions this in https://github.com/rollup/rollup/issues/1828#issuecomment-406630148)

Without modifying rollup itself there are only 2 solutions right now:

1. Disable safe writes in your editor
2. Add usePolling: true to the chokidar config (or equivalent setting for other watchers)

If this were to be fixed in rollup itself I believe we'd need a way to tell it to look at the directory that our files belong in. The issue here is that rather than being able to treeshake the code and know exactly what files to watch you risk watching too much (such as if you reference something in node-modules and want to watch 1 file in the module rather than the entire module itself).

I see 2 potential options:

1. You tell rollup to blindly watch the directory of the files it wants to watch (rather than the files directly).
2. You manually tell rollup what directories to watch and the burden falls on you to know what you need watched.

Example of `usePolling`:
```
rollup.config.js

export default {
    watch: {
        chokidar: {
            usePolling: true
        }
    },
    ...
}
```
Or if you're using code splitting:
```
rollup.config.js

export default {
    client: {
        watch: {
            chokidar: {
                usePolling: true
            }
        },
        ...
    },
    server: {
        watch: {
            chokidar: {
                usePolling: true
            }
        },
        ...
    }
}
```

**Comment by lukastaegert** [comment]
Hi @paco3346, these are important findings that could help us improve here, but since they are in a closed issue, they are likely to go unnoticed. The current plan is to move to only supporting chokidar in the next major. #2988 would be the relevant issue here, I will link them there.

**Comment by getkey** [comment]
@paco3346 thanks a lot for researching this!

To fellow (Neo)Vim users: do [`set backupcopy=yes`](http://vimdoc.sourceforge.net/htmldoc/options.html#'backupcopy') to disable safe writes.

**Comment by vineetit1991** [comment]
I am on rollup@1.26.0 and tried everything for watch mode but nothing is working for me.


```
{
	watch: {
		include: 'src/**/*.ts',
		chokidar: {
		    usePolling: true
		},
		exclude: "node_modules/**"
	},
	input: "src/index.ts",
	output: [{
		file: pkg.main,
		format: "umd",
		sourcemap: true,
		name: 'main'
    	}, {
		file: pkg.module,
		sourcemap: true,
		format: "es"
	}],
	external: [
		...Object.keys(pkg.dependencies || {}),
		...Object.keys(pkg.peerDependencies || {}),
	],
  	plugins: [
		eslint({
			include: ["src/**/*.{ts|tsx}"],
			exclude: ['src/**/*.css']
		}),
		postcss({
			plugins: []
		}),
		typescriptRollup({ typescript }),
		babel({
			exclude: 'node_modules/**',
		}),
		json
	]
}
```

Is there any way to work around this ?

**Comment by JeffWScott** [comment]
Also stops working for me.  I tried all settings in this thread but I cannot get the localhost to show my latest changes.

**Comment by dman777** [comment]
I have this issue also. Just using VIM. Seems ` chokidar: false` fixed it.

**Comment by davinderpalrehal** [comment]
Hey everyone,

I have been using Gulp and Rollup for the past couple of days, was all working fine and today it just stopped working. I have tried all the suggested up here nothings seems to work for me, i end up having to restart Gulp everytime i make a change to JS. Its the same when using Rollup on its on

**Comment by shellscape** [comment]
Consider using `rollup-stream` for use with Gulp.

A side note for everyone participating on this thread: While other users might jump in with tips and help, Rollup doesn't offer support in issues (closed or open). The best thing one can do if facing a problem without a solution is to open a new bug issue - following the issue template to the letter - and include a proper _**minimum viable**_ reproduction. Remember, we can only help if you include a reproduction.

**Comment by davinderpalrehal** [comment]
Thanks @shellscape , will try it out, though I am still baffled as to why it stopped working in the first place. And also on the official Rollup documentation, there is a section for using Rollup with Gulp (https://rollupjs.org/guide/en/#gulp) which should be updated if thats not the recommended way of doing things.

**Comment by shellscape** [comment]
The main docs list _one_ method, but not _the only_ method.

**Comment by davinderpalrehal** [comment]
Thanks everyone, moved to `@rollup/stream` and all is working so far

**Comment by KASOGIT** [comment]
Hi, i'm facing the same issue with this config on macOS 10.15.6
```javascript
// rollup.config.js
import typescript from '@rollup/plugin-typescript';
import commonjs from '@rollup/plugin-commonjs';
import url from '@rollup/plugin-url';
import json from '@rollup/plugin-json';
import peerDepsExternal from 'rollup-plugin-peer-deps-external';

export default {
  input: 'src/index.ts',
  output: {
    dir: 'dist',
    format: 'esm',
    sourcemap: true,
  },
  plugins: [
    peerDepsExternal(),
    commonjs(),
    json(),
    url({
      include: ['**/*.svg', '**/*.png', '**/*.jpg', '**/*.gif', '**/*.m4a'],
      // to-do: export assets to library consummers
      limit: Number.POSITIVE_INFINITY,
    }),
    typescript(),
  ],
};

```
Sometimes for no reason on a file change, a git pull or something similar rollup exit the watch mode with code 0 that's really.
Does someone using a look like config solve this issue ?
